autoreload verbosity

shaperilio opened the pull request:
Worked on three things:
1. More descriptive parameter names for `%autoreload`; `now`, `off`, `explicit`, `all`, `complete`. (This last one could probably use a better name, but I couldn't think of anything better based on the message in 1d3018a93e98ad55f41d4419f835b738de80e1b7)
2. New optional arguments for `%autoreload` allow displaying the names of modules that are reloaded. Use `--print` or `-p` to use `print` statements, or `--log` / `-l` to log at `INFO` level.
3. `%aimport` can parse whitelist/blacklist modules on the same line, e.g. `%aimport os, -math` now works.

`%autoreload` and will also now raise a `ValueError` if the parameter is invalid. I suppose a bit more verification could be done for input to `%aimport`....

shaperilio committed:
Remove copy-paste error

shaperilio committed:
Verbose version of `%autoreload` command

shaperilio committed:
Improve parsing for `%aimport`

shaperilio committed:
Raise error on bad parameter

shaperilio committed:
Add `%averbose` magic command.

shaperilio committed:
Correct documentation for mode 3

shaperilio committed:
Test for verbose mode names

shaperilio committed:
Test for new parsing of `%aimport`

shaperilio committed:
Test of `%averbose`

Cannot get logging output to show in tests, TBD if that's needed

shaperilio committed:
RST file documentation

shaperilio committed:
Fixes from `darker`

shaperilio committed:
doc formatting fixes

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

shaperilio commented:
Can I get some feedback on this?

krassowski commented:
Thank you for working on it. Is there an issue where the proposed enhancements were previously discussed? If so, could you link it? If not, could you search for related issues to summarise any prior/related discussions?

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -595,15 +618,41 @@ def aimport(self, parameter_s="", stream=None):
+                    self.shell.push({top_name: top_module})
+
+    @line_magic
+    def averbose(self, parameter_s=""):
+        r"""%averbose => Turn verbosity on/off for autoreloading.
```
Personally I would prefer if instead of adding another magic, the `%autoreload` magic accepted `--verbose` argument. There is already a mechanism for parsing arguments in magics.

shaperilio replied:
Sounds great; I'll work on this.

shaperilio replied:
This is done; added `--print` and `--log` options for `%autoreload`.

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -595,15 +618,41 @@ def aimport(self, parameter_s="", stream=None):
 
-                # Inject module to user namespace
-                self.shell.push({top_name: top_module})
+        %averbose 0 or %averbose off
```
I would avoid adding more magic numbers. off/on/log is fine.

shaperilio replied:
I'll get rid of `%averbose` as discussed above.

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +573,22 @@ def autoreload(self, parameter_s=""):
 
         """
-        if parameter_s == "":
+        if parameter_s == "" or parameter_s.lower() == "now":
```
You could add a new variable like `parameter_s_lower = parameter_s.lower()` and only use that in tests to save quote a few calls. Even better, this could be solved with a dictionary but probably not in this PR.

shaperilio replied:
Is this a style thing or an efficiency thing? Happy to make the change; I'm just curious.

krassowski reviewed (commented):
Some high-level comments.

shaperilio commented:
> Thank you for working on it. Is there an issue where the proposed enhancements were previously discussed? If so, could you link it? If not, could you search for related issues to summarise any prior/related discussions?

I did not find any relevant discussions when I did a search prior to working on this, though I may have missed something since I'm not active in the repo.

We use `%autoreload` often in one of my projects, and have been "bitten" by some of the quirks (i.e. not realizing we had changed a class and then getting "nonsense" errors after autoreload). So we modified the source so that autoreload shows which modules it's updating.

I found this very useful, so I decided to work on submitting the change upstream.

At that point I realized I could also change the arguments from simple ints to something more descriptive--I've never met anyone in person who knew what they were, I was always told "use 2" or whatever they were passing along from previous experience.

shaperilio committed:
Call `lower` only once

shaperilio committed:
Replace `%averbose` with args to `%autoreload`

shaperilio committed:
Add logging tests

shaperilio commented:
I made changes as you suggested. Let me know if there are any other improvements you'd like to see.

shaperilio committed:
Format with `black`

shaperilio committed:
Reformat test with `black`

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +605,45 @@ def autoreload(self, parameter_s=""):
+            self._reloader._report = lambda msg: None
+        elif args.print is True:
+            if args.log is True:
+                self._reloader._report = lambda msg: pl(msg)
```
Unless there is a reason for wrapping in lambdas I would suggest to just assign the function directly.

```suggestion
                self._reloader._report = pl
```

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +605,45 @@ def autoreload(self, parameter_s=""):
+            if args.log is True:
+                self._reloader._report = lambda msg: pl(msg)
+            else:
+                self._reloader._report = lambda msg: p(msg)
```
```suggestion
                self._reloader._report = p
```

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +605,45 @@ def autoreload(self, parameter_s=""):
+            else:
+                self._reloader._report = lambda msg: p(msg)
+        elif args.log is True:
+            self._reloader._report = lambda msg: l(msg)
```
```suggestion
            self._reloader._report = l
```

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +605,45 @@ def autoreload(self, parameter_s=""):
+        args = magic_arguments.parse_argstring(self.autoreload, line)
+        mode = args.mode.lower()
+
+        def p(msg):
+            print(msg)
```
Worth making it one func def and one call on each reload less?

```suggestion
        p  = print
```

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -552,21 +605,45 @@ def autoreload(self, parameter_s=""):
+        def p(msg):
+            print(msg)
+
+        def l(msg):
+            logging.getLogger("autoreload").info(msg)
```
Unless there is a case for hot-swapping the logger at execution.

```suggestion
        logger = logging.getLogger("autoreload")
        def l(msg):
            logger.info(msg)
```

krassowski commented on IPython/extensions/autoreload.py:
```diff
@@ -581,8 +658,9 @@ def aimport(self, parameter_s="", stream=None):
-        %aimport -foo
-        Mark module 'foo' to not be autoreloaded for %autoreload 1
+        %aimport -foo, bar
+        Mark module 'foo' to not be autoreloaded for %autoreload 1, 2, or 3, and 'bar'
+        to be autoreloaded for 1.
```
Feel welcome to replace the digits with verbose names here and in the paragraph above if you agree that this is a good idea.

shaperilio replied:
Done; thanks for the suggestion.

krassowski commented on docs/source/whatsnew/pr/autoreload-verbosity.rst:
```diff
@@ -0,0 +1,21 @@
+Autoreload verbosity
+====================
+
+We introduce more descriptive names for the `%autoreload` parameter:
```
Two suggestions:
- replace all single backticks with double backtics
- add a new line here so that the list renders nicely

rst has quite specific requirements.

```suggestion
We introduce more descriptive names for the ``%autoreload`` parameter:

```

shaperilio replied:
Sorry about that. I rendered the RST and think I got it right now.

krassowski reviewed (commented):
Tested locally works perfectly. Small suggestions to reduce number of function definitions and RST syntax.

shaperilio committed:
Apply suggestions from code review

Co-authored-by: Michał Krassowski <5832902+krassowski@users.noreply.github.com>

shaperilio committed:
Fix RST syntax

shaperilio committed:
Use new names in docstring.

shaperilio committed:
`l` doesn't need to be a function

shaperilio committed:
Fix formatting

shaperilio committed:
Fully undo mode flags

krassowski reviewed (approved):
Thanks!

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

shaperilio commented:
I'm not sure what to do about this failing doc build...

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

krassowski commented:
The docs build will be fixed by https://github.com/ipython/ipython/pull/13872. Thanks for highlighting it.

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

shaperilio commented:
@krassowski I don't know what the failed test is about, is there a way to rerun it?

krassowski commented:
Yes, you can close and then immediately reopen the pull request to kick GitHub actions.

shaperilio commented:
Cool, that worked. Everything is passing now.

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

shaperilio committed:
Merge branch 'main' into shaperilio/autoreload-verbosity

krassowski commented:
@fperez what do you think about this one? This would also fix https://github.com/ipython/ipython/issues/13915 (as discussed in https://github.com/ipython/ipython/pull/13916).

shaperilio commented:
FYI I updated the original comment at the top of the PR to reflect the behavior after all your suggestions.

shaperilio commented:
@Carreau @fperez can we merge this one for 8.11 too?

fperez reviewed (approved):
Excellent, I very much like these improvements, many thanks! They'll come in handy in my current class.

Tagging @facusapienza21 and @isaacdsloan for awareness, we'll be using this when released.

fperez commented:
Done @shaperilio, many thanks for the great improvements and well documented/tested work. Much appreciated (and sorry for the slow response!).
